// aws-cdk/test/api/deploy-stack.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lib_1 = require("../../lib");
const util_1 = require("../util");
const mock_sdk_1 = require("../util/mock-sdk");
const FAKE_STACK = util_1.testStack({
    stackName: 'withouterrors',
    template: { resource: 'noerrorresource' },
});
test('do deploy executable change set with 0 changes', async () => {
    // GIVEN
    const sdk = new mock_sdk_1.MockSDK();
    let executed = false;
    sdk.stubCloudFormation({
        describeStacks() {
            return {
                Stacks: []
            };
        },
        createChangeSet() {
            return {};
        },
        describeChangeSet() {
            return {
                Status: 'CREATE_COMPLETE',
                Changes: [],
            };
        },
        executeChangeSet() {
            executed = true;
            return {};
        }
    });
    // WHEN
    const ret = await lib_1.deployStack({
        stack: FAKE_STACK,
        sdk,
    });
    // THEN
    expect(ret.noOp).toBeFalsy();
    expect(executed).toBeTruthy();
});
