// aws-cdk/test/context-providers/amis.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const aws = require("aws-sdk");
const AWS = require("aws-sdk-mock");
const ami_1 = require("../../lib/context-providers/ami");
AWS.setSDKInstance(aws);
afterEach(done => {
    AWS.restore();
    done();
});
const mockSDK = {
    defaultAccount: () => Promise.resolve('123456789012'),
    defaultRegion: () => Promise.resolve('bermuda-triangle-1337'),
    cloudFormation: () => { throw new Error('Not Mocked'); },
    ec2: () => Promise.resolve(new aws.EC2()),
    ecr: () => { throw new Error('Not Mocked'); },
    route53: () => { throw new Error('Not Mocked'); },
    s3: () => { throw new Error('Not Mocked'); },
    ssm: () => { throw new Error('Not Mocked'); },
};
test('calls DescribeImages on the request', async () => {
    // GIVEN
    let request;
    AWS.mock('EC2', 'describeImages', (params, cb) => {
        request = params;
        return cb(null, { Images: [{ ImageId: 'ami-1234' }] });
    });
    // WHEN
    await new ami_1.AmiContextProviderPlugin(mockSDK).getValue({
        account: '1234',
        region: 'asdf',
        owners: ['some-owner'],
        filters: {
            'some-filter': ['filtered']
        }
    });
    // THEN
    expect(request).toEqual({
        Owners: ['some-owner'],
        Filters: [
            {
                Name: 'some-filter',
                Values: ['filtered'],
            }
        ]
    });
});
test('returns the most recent AMI matching the criteria', async () => {
    // GIVEN
    AWS.mock('EC2', 'describeImages', (_, cb) => {
        return cb(null, {
            Images: [
                {
                    ImageId: 'ami-1234',
                    CreationDate: "2016-06-22T08:39:59.000Z",
                },
                {
                    ImageId: 'ami-5678',
                    CreationDate: "2019-06-22T08:39:59.000Z",
                }
            ]
        });
    });
    // WHEN
    const result = await new ami_1.AmiContextProviderPlugin(mockSDK).getValue({
        account: '1234',
        region: 'asdf',
        filters: {}
    });
    // THEN
    expect(result).toBe('ami-5678');
});
