// aws-cdk/test/diff.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cxapi = require("@aws-cdk/cx-api");
const stream_1 = require("stream");
const string_decoder_1 = require("string_decoder");
const stacks_1 = require("../lib/api/cxapp/stacks");
const sdk_1 = require("../lib/api/util/sdk");
const cdk_toolkit_1 = require("../lib/cdk-toolkit");
const settings_1 = require("../lib/settings");
const util_1 = require("./util");
const FIXED_RESULT = util_1.testAssembly({
    stacks: [{
            stackName: 'A',
            template: { resource: 'A' },
        },
        {
            stackName: 'B',
            depends: ['A'],
            template: { resource: 'B' },
        },
        {
            stackName: 'C',
            depends: ['A'],
            template: { resource: 'C' },
            metadata: {
                '/resource': [
                    {
                        type: cxapi.ERROR_METADATA_KEY,
                        data: 'this is an error'
                    }
                ]
            }
        }]
});
const appStacks = new stacks_1.AppStacks({
    configuration: new settings_1.Configuration(),
    aws: new sdk_1.SDK(),
    synthesizer: async () => FIXED_RESULT,
});
test('diff can diff multiple stacks', async () => {
    // GIVEN
    const provisioner = {
        async readCurrentTemplate(_stack) {
            return {};
        },
        async deployStack(_options) {
            return { noOp: true, outputs: {}, stackArn: '' };
        }
    };
    const toolkit = new cdk_toolkit_1.CdkToolkit({ appStacks, provisioner });
    const buffer = new StringWritable();
    // WHEN
    const exitCode = await toolkit.diff({
        stackNames: ['B'],
        stream: buffer
    });
    // THEN
    const plainTextOutput = buffer.data.replace(/\x1B\[[0-?]*[ -/]*[@-~]/g, '');
    expect(plainTextOutput).toContain('Stack A');
    expect(plainTextOutput).toContain('Stack B');
    expect(exitCode).toBe(0);
});
test('exits with 1 with diffs and fail set to true', async () => {
    // GIVEN
    const provisioner = {
        async readCurrentTemplate(_stack) {
            return {};
        },
        async deployStack(_options) {
            return { noOp: true, outputs: {}, stackArn: '' };
        }
    };
    const toolkit = new cdk_toolkit_1.CdkToolkit({ appStacks, provisioner });
    const buffer = new StringWritable();
    // WHEN
    const exitCode = await toolkit.diff({
        stackNames: ['A'],
        stream: buffer,
        fail: true
    });
    // THEN
    expect(exitCode).toBe(1);
});
test('throws an error during diffs on stack with error metadata', async () => {
    // GIVEN
    const provisioner = {
        async readCurrentTemplate(_stack) {
            return {};
        },
        async deployStack(_options) {
            return { noOp: true, outputs: {}, stackArn: '' };
        }
    };
    const toolkit = new cdk_toolkit_1.CdkToolkit({ appStacks, provisioner });
    const buffer = new StringWritable();
    // WHEN
    try {
        const exitCode = await toolkit.diff({
            stackNames: ['C'],
            stream: buffer
        });
        // THEN
        expect(exitCode).toBe(1);
    }
    catch (e) {
        expect(e.toString()).toContain('Found errors');
    }
});
class StringWritable extends stream_1.Writable {
    constructor(options = {}) {
        super(options);
        this._decoder = new string_decoder_1.StringDecoder(options && options.defaultEncoding);
        this.data = '';
    }
    _write(chunk, encoding, callback) {
        if (encoding === 'buffer') {
            chunk = this._decoder.write(chunk);
        }
        this.data += chunk;
        callback();
    }
    _final(callback) {
        this.data += this._decoder.end();
        callback();
    }
}
